From 4241bbf0e471ac2e0d1c4491c340c944a318913e Mon Sep 17 00:00:00 2001
From: Avula Sri Charan <quic_asrichar@quicinc.com>
Date: Sun, 5 Jan 2025 11:14:19 +0530
Subject: UPSTREAM:CP:wifi:ath12k: Skip assigning channel context if radio is not active

In case of multiple countries mentioned in the wireless config file the
regulatory rules will be updated for every different country. But  if
in between of these events if there's radio_start request from mac80211
the radio will comply with the regulatory rules at that point of time
which might sometimes lead to bringing down a radio.

Post this if there's a new regdb available where some of the previously
disabled radio's are valid here, mac80211 will try to create and start
vdev on a radio that is not even active with respect to the driver.

To avoid this add an additional check in ath12k_get_ar_by_ctx() to check
if the radio is active or not and return NULL if it's notactive. This
will ensure that any request from mac80211 on in active radio will be
discarded.

Patch-dependency:
	652-02-ath12k-Update-hw-generic-mac-ops-handling.patch

Signed-off-by: Avula Sri Charan <quic_asrichar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -747,7 +747,7 @@ struct ath12k *ath12k_get_ar_by_link_idx
 	ar = ah->radio;
 	for (int i = 0; i < ah->num_radio; i++) {
 		if (ar->link_idx == link_idx)
-			return ar;
+			return ath12k_mac_get_ar_by_pdev_id(ar->ab, ar->pdev->pdev_id);
 		ar++;
 	}
 	return NULL;
@@ -1098,10 +1098,17 @@ static struct ath12k *ath12k_mac_get_ar_
 static struct ath12k *ath12k_get_ar_by_ctx(struct ieee80211_hw *hw,
 					   struct ieee80211_chanctx_conf *ctx)
 {
+	struct ath12k *ar;
+
 	if (!ctx)
 		return NULL;
 
-	return ath12k_mac_get_ar_by_chan(hw, ctx->def.chan);
+	ar = ath12k_mac_get_ar_by_chan(hw, ctx->def.chan);
+
+	if (!ar || ath12k_mac_get_ar_by_pdev_id(ar->ab, ar->pdev->pdev_id))
+		return ar;
+
+	return NULL;
 }
 
 struct ath12k *ath12k_get_ar_by_vif(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
@@ -6141,9 +6148,9 @@ ath12k_mac_select_scan_device(struct iee
 			if (center_freq > ar->chan_info.low_freq &&
 			    center_freq < ar->chan_info.high_freq)
 				if (ar->mac.sbands[band].channels)
-					return ar;
+					return ath12k_mac_get_ar_by_pdev_id(ar->ab, ar->pdev->pdev_id);
 		} else if (ar->mac.sbands[band].channels) {
-			return ar;
+			return ath12k_mac_get_ar_by_pdev_id(ar->ab, ar->pdev->pdev_id);
 		}
 		ar++;
 	}
@@ -16432,8 +16439,11 @@ ath12k_mac_assign_vif_chanctx_handle(str
 		ar = ath12k_mac_assign_vif_to_vdev(hw, arvif, ctx,
 						   is_bridge_vdev,
 						   bridge_ar_link_idx);
-		if (!ar)
+		if (!ar) {
+			ath12k_mac_remove_link_interface(hw, arvif);
+			ath12k_mac_unassign_link_vif(arvif);
 			return -EINVAL;
+		}
 
 		ath12k_rm_link_state_update(ar->pdev_idx,
 					    ar->ab, arvif,
